rust/noosphere-ns/src/bin/orb-ns/runner/runner.rs

```diff
@@ -0,0 +1,145 @@
+        let peer_id = node.peer_id().to_owned();
+
+        let listening_address = if let Some(requested_addr) = config.listening_address.take() {
+            // Request address from DHT to resolve default port (0) to
+            // selected port.
+            let resolved_addr = node.listen(requested_addr.to_owned()).await?;
+            Some(resolved_addr)
+        } else {
+            None
+        };
+
+        node.add_peers(config.peers.to_owned()).await?;
+
+        let wrapped_node = Arc::new(Mutex::new(node));
+
+        let (api_address, api_thread) = if cfg!(feature = "api-server") {
+            if let Some(requested_addr) = config.api_address.take() {
+                let api_listener = TcpListener::bind(requested_addr)?;
+                let api_address = socket_addr_to_url(api_listener.local_addr()?)?;
+                (
+                    Some(api_address),
+                    Some(APIServer::serve(wrapped_node.clone(), api_listener)),
+                )
+            } else {
+                (None, None)
+            }
+        } else {
+            (None, None)
+        };
+
+        Ok(ActiveNameSystem {
+            name_system: wrapped_node,
+            peer_id,
+            listening_address,
+            api_address,
+            _api_thread: api_thread,
+        })
+    }
+
+    pub(crate) async fn bootstrap_peers(&self, addresses: &Vec<Multiaddr>) -> Result<()> {
+        let node = self.name_system.lock().await;
+
+        if !addresses.is_empty() {
+            // Add both local peers (other nodes hosted in this process)
+            // and provided bootstrap peers
+            node.add_peers(addresses.clone()).await?;
+        }
+
+        node.bootstrap().await?;
+        Ok(())
+    }
+}
+
+/// Holds several running name systems.
+#[derive(Serialize)]
+pub struct NameSystemRunner {
+    name_systems: Vec<ActiveNameSystem>,
+}
+
+/// Runner runs one or many DHT nodes based off of provided
+/// configuration from a [CLICommand].
+impl NameSystemRunner {
+    pub async fn try_from_config(config: RunnerConfig) -> Result<Self> {
+        let ns_iter: Vec<_> = config
+            .nodes
+            .into_iter()
+            .map(ActiveNameSystem::try_from_config)
+            .collect();
+        let name_systems: Vec<ActiveNameSystem> = try_join_all(ns_iter).await?;
+        let addresses: Vec<Multiaddr> = name_systems
+            .iter()
+            .filter_map(|r| r.listening_address.to_owned())
+            .collect();
+
+        for ns in name_systems.iter() {
+            ns.bootstrap_peers(&addresses).await?;
+        }
+
+        Ok(NameSystemRunner { name_systems })
+    }
+}
+
+/// Future implementation for [NameSystemManager] so we can
+/// keep alive on a pending future for the necessary resources.
+impl Future for NameSystemRunner {
+    type Output = Result<()>;
+
+    fn poll(self: Pin<&mut Self>, _cx: &mut task::Context<'_>) -> task::Poll<Self::Output> {
+        task::Poll::Pending
```

Comment:
Suggestions on better ways of leaving the main process alive?

Reply:
Can't `NameSystemRunner` just vend a future that awaits on whatever is long-running?

Reply:
Deferring to the shell's ctrl-c handling, the lifetime of the NameServer in the `run` command matches the process